### DISCUSSION TITLE
pydantic object __init()__ function call revealed in fastAPI type_error payload

### DISCUSSION THREAD

**Discussion by Syffe** [fastapi/fastapi#8624] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import pydantic

app = FastAPI()

@pydantic.dataclasses.dataclass
class TestPayload:
    message: str = dataclasses.field(
        metadata={"description": "test message"}
    )

@app.put('/test_endpoint')
def create(test_payload: TestPayload) -> TestPayload:
    return test_payload


payload = {"false_key": "test message"}

app.put("/test_endpoint",
        json=payload,
      )
```

### Description

Hi, 

When trying to use a pydantic model as a parameter for an endpoint, if this said payload isn't matching the pydantic model defined, with, in this case, a wrong key, fastAPI will return an error payload as follows:

```python
# fastAPI error payload returned
{
'loc': ('body',), 
'msg': "TestPayload.__init__() got an unexpected keyword argument 'false_key'", 
'type': 'type_error'
}
```

The error message reveals that .__init()__ is called from the pydantic object, and this kind of message is not quiet what an API user would like to see following a key error. 
The question then being, Is it the good behavior following those kind of errors ? Is it ok to reveal this kind of call information ? 
May be it is a bug, or an error that isn't correctly catched on the server side

### Operating System

macOS

### FastAPI Version

0.74.1

### Python Version

Python 3.10.2

**Comment by harunyasar** [comment]
If you ask me pydantic dataclass is not directly related to FastAPI's API. I mean when you pass `{"false_key": "test message"}`, it will be interpreted like `def __init__(message: str)` by dataclass. So, in your example, you are simply passing like as follows.

```python
test_payload = TestPayload(false_key="test message")
```

This means you are passing an argument with the wrong name to the initialization method.

**Comment by sileht** [comment]
The issue here is that FastAPI returns a different error if `pydantic.dataclasses` is used instead `pydantic.BaseModel`.

pydantic.BaseModel returns:
```
[
     {'loc': ['body', 'reason'], 'msg': 'field required', 'type': 'value_error.missing'}
]
```
while dataclasses variant returns: 
```
[{
    'loc': ('body',), 
    'msg': "TestPayload.__init__() got an unexpected keyword argument 'false_key'", 
    'type': 'type_error'
}]
```

The dataclasses versions leak implementation details, which is a bug for me as dataclasses variant is officially supported: https://fastapi.tiangolo.com/advanced/dataclasses/

**Comment by sileht** [comment, marked as answer]
Looks like the bug has been fixed, the error message is the same now.
